placement/placement.go

```diff
@@ -0,0 +1,667 @@
+		shardCutovers[name] = append(shardCutovers[name], added...)
+		assigned.Shards = append(assigned.Shards, added...)
+
+		// Sort the shards so if we unassign later, the results will be deterministic
+		sort.Sort(shardsInOrder(assigned.Shards))
+	}
+
+	// now build out all of the rules
+	db.Version++
+	rules := &schema.ClusterMappingRuleSet{
+		ForVersion: db.Version,
+	}
+
+	var (
+		readCutoverTime  = sp.clock.Now()
+		writeCutoverTime = readCutoverTime.Add(opts.GetRolloutDelay())
+		cutoffTime       = writeCutoverTime.Add(opts.GetTransitionDelay())
+	)
+
+	for name, shards := range shardCutovers {
+		sp.log.Infof("Shifting %d shards onto %s:%s at (reads starting @ %v, writes starting @ %v)",
+			len(shards), db.Name, name, readCutoverTime, writeCutoverTime)
+
+		rules.Cutovers = append(rules.Cutovers, &schema.CutoverRule{
+			ClusterName:      name,
+			Shards:           shards,
+			ReadCutoverTime:  xtime.ToUnixMillis(readCutoverTime),
+			WriteCutoverTime: xtime.ToUnixMillis(writeCutoverTime),
+		})
+	}
+
+	for name, shards := range shardCutoffs {
+		sp.log.Infof("Shifting %d shards off %s:%s (reads stopping @ %v)",
+			len(shards), db.Name, name, cutoffTime)
+
+		rules.Cutoffs = append(rules.Cutoffs, &schema.CutoffRule{
+			ClusterName: name,
+			Shards:      shards,
+			CutoffTime:  xtime.ToUnixMillis(cutoffTime),
+		})
+
+	}
+
+	// Sort rules by cluster name so that they have a deterministic ordering...
+	sort.Sort(cutoffsByCluster(rules.Cutoffs))
+	sort.Sort(cutoversByCluster(rules.Cutovers))
+
+	// ...and add to the list of mappings for the database
+	db.MappingRules = append(db.MappingRules, rules)
+	return nil
+}
+
+// computeDesiredNumShards computes the desired number of shards per cluster,
+// based on the cluster weight and total number of shards
+func (sp storagePlacement) computeDesiredNumShards(
+	clusters map[string]*schema.Cluster,
+	totalShards int) map[string]int {
+
+	if len(clusters) == 0 {
+		return nil
+	}
+
+	// Compute the total weight of all clusters
+	var totalWeight uint32
+	for _, c := range clusters {
+		totalWeight += c.Weight
+	}
+
+	// compute desired # of shards based on relative weight of each cluster
+	var (
+		oldestCluster    *schema.Cluster
+		shardsRemaining  = totalShards
+		desiredNumShards = make(map[string]int, len(clusters))
+	)
+	for name, c := range clusters {
+		relativeWeight := (float64(c.Weight) / float64(totalWeight))
+		numShards := int(relativeWeight * float64(totalShards))
+		if numShards > shardsRemaining {
+			numShards = shardsRemaining
+		}
+
+		desiredNumShards[name] = numShards
+		shardsRemaining -= numShards
+
+		if oldestCluster == nil || oldestCluster.CreatedAt > c.CreatedAt {
+			oldestCluster = c
+		}
+	}
+
+	// if we haven't yet allocated all shards due to rounding, pick oldest
+	// cluster and add the remainder to it. we pick the oldest cluster to have a
+	// deterministic mapping and avoid moving shards around unnecessarily
+	if shardsRemaining > 0 {
```

Comment:
True, so that could happen if the shards assigned to the last cluster were taken up by other clusters because we are rounding up. On the other hand, the current algorithm may cause the last cluster to take a lot more shards than others because we are rounding down. We could run both and pick the placement that's more balanced but not sure if this happens frequently enough to justify the implementation complexity.